src/main.rs

```diff
@@ -204,3 +174,150 @@ fn parser<'a>() -> impl Parser<'a, &'a str, Vec<Statement<'a>>, extra::Err<Rich<
 }
+
+/// Parses a single value until a space or newline is encountered.
+fn value<'s>(i: &mut &'s str) -> PResult<&'s str> {
+    take_till(0.., |c| c == ' ' || c == '\n').parse_next(i)
+}
+
+/// Parses a single value until a space or newline is encountered
+///
+/// This parser also consumes any trailing whitespace or newlines.
+fn value_terminated<'s>(i: &mut &'s str) -> PResult<&'s str> {
+    (value, alt((space0, line_ending)))
+        .parse_next(i)
+        .map(|(v, _)| v)
+}
+
+/// Parses the initial `#mtree` line at the top of the file.
+fn init<'s>(i: &mut &'s str) -> PResult<Statement<'s>> {
+    terminated("#mtree", line_ending)
+        .parse_next(i)
+        .map(|_| Statement::Init)
+}
+
+/// Parses a single default property.
+fn default_property<'s>(i: &mut &'s str) -> PResult<DefaultProperty<'s>> {
+    separated_pair(alpha1, "=", value_terminated)
+        .verify_map(|(k, v)| match k {
+            "uid" => Some(DefaultProperty::Uid(v.parse().ok()?)),
+            "gid" => Some(DefaultProperty::Gid(v.parse().ok()?)),
+            "type" => match v {
+                "dir" => Some(DefaultProperty::Type(PathType::Dir)),
+                "file" => Some(DefaultProperty::Type(PathType::File)),
+                "link" => Some(DefaultProperty::Type(PathType::Link)),
+                _ => panic!("unknown type: {v}"),
+            },
+            "mode" => Some(DefaultProperty::Mode(v)),
+            _ => panic!("unknown property: {k}"),
+        })
+        .parse_next(i)
+}
+
+/// Parses a list of default properties.
+fn default_properties<'s>(i: &mut &'s str) -> PResult<Vec<DefaultProperty<'s>>> {
+    repeat(0.., default_property).parse_next(i)
+}
+
+/// Parses a `/set` command followed by some properties.
+fn set<'s>(i: &mut &'s str) -> PResult<Statement<'s>> {
+    delimited(("/set", space0), default_properties, line_ending)
+        .parse_next(i)
+        .map(Statement::Set)
+}
+
+/// Parses a `/unset` command followed by some properties.
+fn unset<'s>(i: &mut &'s str) -> PResult<Statement<'s>> {
+    delimited(("/unset", space0), default_properties, line_ending)
+        .parse_next(i)
+        .map(Statement::Unset)
+}
+
+/// Parses a single property.
+fn property<'s>(i: &mut &'s str) -> PResult<Property<'s>> {
+    separated_pair(alphanumeric1, "=", value_terminated)
+        .verify_map(|(k, v)| match k {
+            "type" => match v {
+                "dir" => Some(Property::Type(PathType::Dir)),
+                "file" => Some(Property::Type(PathType::File)),
+                "link" => Some(Property::Type(PathType::Link)),
+                _ => panic!("unknown type: {v}"),
+            },
+            "mode" => Some(Property::Mode(v)),
+            "size" => Some(Property::Size(v.parse().ok()?)),
+            "link" => Some(Property::Link(v)),
+            "sha256digest" => Some(Property::Sha256Digest(v)),
+            "time" => Some(Property::Time(v.split_once(".")?.0.parse().ok()?)),
```

Comment:
Not sure how to refactor this. I tried the following but gave up half way through because I didn't like where it was going:

```rust
digit1.parse_to().parse_next(&mut v.split_once(".")?.0)?,
```

Reply:
https://gitlab.archlinux.org/archlinux/alpm/alpm/-/commit/c0e3b67808b7e32b6e230b9f2c58bc3549f764f0#d29a4370b2ae5708cb954c4f2a9d076e47894108_0_227